WalletWasabi/WabiSabi/Client/ArenaClientResponse.cs

```diff
@@ -0,0 +1,28 @@
+using System.Collections.Generic;
+using System.Linq;
+using WalletWasabi.Crypto.ZeroKnowledge;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class ArenaResponse
+	{
+		public ArenaResponse(IEnumerable<Credential> realAmountCredentials, IEnumerable<Credential> realVsizeCredentials)
+		{
+			RealAmountCredentials = realAmountCredentials.ToArray();
+			RealVsizeCredentials = realVsizeCredentials.ToArray();
```

Comment:
Even though it's private, here and in general (especially the methods that return credentials) I would prefer `ToImmutableArray()` since the mutatability of `Credential[]` is not used:

1. array usage syntax unchanged when using the arrays (linq or normal array access)
2. the change in the type signature conveys guarantees and helps readability especially for future maintainers, because seeing that something is immutable based on its type in another region of the code restricts what can be accidentally done to it and informs the reader that other parts of the code can't affect it for the duration of the current scope).

Reply:
We are not really using immutable. I think it is obvious that the array should not be changed. I think this would be a coding convention change, that should be discussed with the rest of the team. So I would live this as it is.

Reply:
I understand you both. In this particular case it serves only as a way to communicate the fact that those arrays are not modified.

I think c# 9 has been a big step forward in many aspects and we should embrace many of those like _record_ and _immutable_ data structures (after all a record is not really immutable if it contains a list or array, for example).  One of the places where the usage of immutable data structures should be less controversial is in situations like this one where they make no difference in usage.

Reply:
@molnard let's continue [this thread](https://github.com/zkSNACKs/WalletWasabi/pull/5726#discussion_r627324102) here?

IMO C# does a great job of making switching between immutable and immutable basically costless from a syntax point of view, read only usage is the same and `var` will do the right thing if the type changes from under it, so using immutable types where mutability is not required seems like the same kind of win as making nullability explicit.

I don't think the coding convention change argument works for or against the types themselves, rather convention determines more which approach should be taken: an approach that uses mutation or one that avoids it. If coding a specific feature relyingon mutation is fine based on conventions and is simpler than one that avoids immutability, then it probably makes sense to do it. If both approaches are both simple, the immutable one is almost always better because it is easier to understand and debug, especially in a concurrent setting.

In this case it appears that not mutating the credentials is in line with conventions, and a nice promise for this API to make (`IEnumerable<Credential>` would provide the same promise, FWIW,).

Separately from which approach to mutation is preferred, assuming that an immutable approach is in use I think it's always strictly a win to also tell the compiler and the reader that that's what that code is doing, since it's almost costless to change the types to allow mutation, and needing to change a bit of boillerplate is a clear signal to anybody working on the code later that a quick fix utilizing mutation might break some assumptions made by other users of the code.

IMO `ImmutableList<T>` and `ImmutableArray<T>` are just useful interfaces that fit somewhere in between `IEnumerable<T>` and `List<T>` or `T[]`, they make more guarantees than `IEnumerable` (including stronger guarantees about repeated evaluation and thread/concurrency safety) but they provide a narrower API than the full mutable APIs. If it makes to use `IEnumerable<T>` over `T[]` in places where it should be more general, then it also makes sense to use the immutable variants where `IEnumerable<T>` is too general, but `T[]` or `List<T>` is still too powerful and requires extra steps to protect from public APIs.